src/profiler/profiler.jl

```diff
@@ -1,74 +1,90 @@
 module Profiler
 
-using Profile, JSON
-
-using JSON
-using Lazy, Juno, Hiccup
-import Juno: Row, LazyTree, SubTree, icon
-import ..Atom: baselink, cliptrace, expandpath, @msg, handle
-using Base.StackTraces
-
-include("tree.jl")
-
-function traces()
-  traces, stacks = Profile.flatten(Profile.retrieve()...)
-  @>>(split(traces, 0, keep = false),
-      map(trace -> @>> trace map(x->stacks[x]) reverse),
-      map(trace -> filter(x->!x.from_c, trace)),
-      filter(x->!isempty(x)))
+using Profile, JSON, FlameGraphs
+
+import ..Atom:  expandpath, @msg, handle
+
+function tojson(node::FlameGraphs.Node, root = false)
+  name, path = expandpath(string(node.data.sf.file))
+  classes = []
+
+  node.data.status & 0x01 ≠ 0 && push!(classes, "isdispatch")
+  node.data.status & 0x02 ≠ 0 && push!(classes, "isgc")
+
+  Dict(
+    :path => path,
+    :location => name,
+    :func => node.data.sf.func,
+    :line => node.data.sf.line,
+    :count => root ? sum(length(c.data.span) for c in node) : length(node.data.span),
+    :classes => classes,
+    :children => [tojson(c) for c in node]
+  )
 end
 
-const NULLFRAME = StackFrame(Symbol(""), Symbol(""), -1)
-
-struct ProfileFrame
-  frame::StackFrame
-  count::Int
+function profiler(data = Profile.fetch() ;kwargs...)
+  graph = FlameGraphs.flamegraph(data ;kwargs...)
+  graph === nothing && return
+  pruneinternal!(graph)
+  prunetask!(graph)
+  @msg profile(tojson(graph, true))
 end
 
-ProfileFrame(frame::StackFrame) = ProfileFrame(frame, 1)
-
-const ProfileTree = Tree{ProfileFrame}
-
-tobranch(trace::StackTrace) = Tree(ProfileFrame(NULLFRAME), [branch(ProfileFrame.(trace))])
-
-mergetrace!(a, b) = merge!(a, b,
-                           (==) = (a, b) -> a.frame == b.frame,
-                           merge = (a, b) -> ProfileFrame(a.frame, a.count + b.count))
-
-rawtree()::ProfileTree = reduce(mergetrace!, tobranch.(traces()))
-
-function cleantree(tree::ProfileTree)
-  postwalk(tree) do x
-    length(x.children) == 1 ? Tree(x.head, x.children[1].children) : x
+function prunetask!(node)
+  for c in node
+    if (
+          (
+            c.data.sf.func == :task_done_hook &&
+            endswith(string(c.data.sf.file), "task.jl")
+          ) ||
+          (
+            # REPL
+            c.data.sf.func == :eval &&
+            FlameGraphs.isleaf(c) &&
+            endswith(string(c.data.sf.file), "boot.jl")
+          )
```

Comment:
this mayn't be so important, but why do we need this ?
Doesn't `pruneinternal!` cut off the branch including `Core.eval` inside of `repleval` anyway ?

Reply:
No, it doesn't. I *think* this is an artifact of all the `@async` stuff happening in Juno, but not sure.